src/runtime_src/core/pcie/tools/xbmgmt/xspi.cpp

```diff
@@ -2180,3 +2224,127 @@ int XSPI_Flasher::upgradeFirmware2Drv(std::istream& mcsStream0,
     // Disable bitstream guard.
     return removeBitstreamGuard(mDev.get(), mFlashDev, bsGuardAddr);
 }
+
+const std::string magic = "XRTDATA";
+struct flash_data_header {
+    char magic[7] = "";
+    char ver = 0;
+    uint32_t data_len = 0;
+    uint32_t parity = 0;
+};
+// Max file data length can be saved on flash
+static const size_t xrt_max_data_len = 1024 * 1024;
+
+static size_t getFlashSize(pcidev::pci_device* dev)
+{
+    std::string err;
+    size_t size;
+
+    dev->sysfs_get<size_t>("flash", "size", err, size, 0);
+    if (size == 0)
+        std::cout << "can't detect flash size: " << err << std::endl;
+    return size;
+}
+
+static uint32_t getParity32(std::vector<unsigned char>& data)
+{
+    uint32_t parity = 0;
+
+    for (size_t len = 0; len < data.size(); len += 4) {
+        uint32_t tmp;
+        std::memcpy(&tmp, data.data() + len, std::min(4ul, data.size() - len));
+        parity ^= tmp;
+    }
+    return parity;
+}
+
+int XSPI_Flasher::xclWriteData(std::vector<unsigned char>& data)
+{
+    const int flash_index = 0;
+
+    if (mFlashDev == nullptr)
+        return -EINVAL;
+
+    if (data.size() > xrt_max_data_len) {
+        std::cout << "meta data is too big (" << data.size() << "B) to flash: "
+            << std::endl;
+        return -EINVAL;
+    }
+
+    size_t size = getFlashSize(mDev.get());
+    if (size == 0)
+        return -EINVAL;
+
+    // Write file data onto flash
+    unsigned int addr = size - sizeof(flash_data_header) - data.size();
+    int ret = writeToFlash(mFlashDev, flash_index, addr,
+        data.data(), data.size());
+    if (ret) {
+        std::cout << "failed to write meta data to flash: " << ret << std::endl;
+        return ret;
+    }
+
+    // Write meta data onto flash
+    flash_data_header header;
+    std::memcpy(header.magic, magic.c_str(), magic.size());
+    header.ver = 0;
+    header.data_len = data.size();
+    header.parity = getParity32(data);
+    addr += data.size();
+    ret = writeToFlash(mFlashDev, flash_index, addr,
+        reinterpret_cast<unsigned char *>(&header), sizeof(header));
+    if (ret) {
+        std::cout << "failed to write meata data header to flash: " << ret
+            << std::endl;
+        return ret;
+    }
+
+    std::cout << "Persisted " << data.size() << " bytes of meta data to flash "
```

Comment:
Being that this code is C++, you might wish to consider using the boost format, which will make your life easier instead of formatting though the iostreams.  For example, the error message becomes:
`std::cout << boost::format("Persisted %d bytes of meta data to flash %d @0x%x\n") % data.size() % flash_index %addr;`

Reply:
will do.